### DISCUSSION TITLE
usb.core.USBError: using PyUSB  for DEM202 Decibel Meter Connected to Laptop

### DISCUSSION THREAD

**Discussion by ckl67** [pyusb/pyusb#437] [Q&A] [answered]
Hello,

I am having trouble trying to obtain data from a DEM202 Velleman sound level meter using PyUSB. 

Laptop OS is Windows 11

	python.exe -m pip list
	...
	pysub               0.2
	pyusb               1.2.1
	..
	
	
	Python version 3.10.5. 

I have used Zadig to replace the drivers with libUSB-win32. 
[See config](https://photos.app.goo.gl/BTNXxsqYnz4wiznV7)

Running the code, There is a usb.core error
I have been stuck some days 
Thank you in advance for your help 

Regards
Christian

```
import time
import usb.core

dev = usb.core.find(idVendor=0x10C4, idProduct=0xEA60)
assert dev is not None
print (dev)

dev.set_configuration()

while True:
    time.sleep(1)
    ret = dev.read(0x82, 200, 100)  
    #ret = dev.ctrl_transfer(0xC1, 8, 0, 0, 200)
    print(ret)

```

Result

```
DEVICE ID 10c4:ea60 on Bus 000 Address 001 =================
 bLength                :   0x12 (18 bytes)
 bDescriptorType        :    0x1 Device
 bcdUSB                 :  0x200 USB 2.0
 bDeviceClass           :    0x0 Specified at interface
 bDeviceSubClass        :    0x0
 bDeviceProtocol        :    0x0
 bMaxPacketSize0        :   0x40 (64 bytes)
 idVendor               : 0x10c4
 idProduct              : 0xea60
 bcdDevice              :  0x100 Device 1.0
 iManufacturer          :    0x1 Silicon Labs
 iProduct               :    0x2 CP2104 USB to UART Bridge Controller
 iSerialNumber          :    0x3 01DA0D3D
 bNumConfigurations     :    0x1
  CONFIGURATION 1: 100 mA ==================================
   bLength              :    0x9 (9 bytes)
   bDescriptorType      :    0x2 Configuration
   wTotalLength         :   0x20 (32 bytes)
   bNumInterfaces       :    0x1
   bConfigurationValue  :    0x1
   iConfiguration       :    0x0
   bmAttributes         :   0x80 Bus Powered
   bMaxPower            :   0x32 (100 mA)
    INTERFACE 0: Vendor Specific ===========================
     bLength            :    0x9 (9 bytes)
     bDescriptorType    :    0x4 Interface
     bInterfaceNumber   :    0x0
     bAlternateSetting  :    0x0
     bNumEndpoints      :    0x2
     bInterfaceClass    :   0xff Vendor Specific
     bInterfaceSubClass :    0x0
     bInterfaceProtocol :    0x0
     iInterface         :    0x2 CP2104 USB to UART Bridge Controller
      ENDPOINT 0x1: Bulk OUT ===============================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :    0x1 OUT
       bmAttributes     :    0x2 Bulk
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0x0
      ENDPOINT 0x82: Bulk IN ===============================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :   0x82 IN
       bmAttributes     :    0x2 Bulk
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0x0
Traceback (most recent call last):
  File "C:\Users\kluges1\Downloads\rasodalo\hello.py", line 12, in <module>
    ret = dev.read(0x82, 200, 100)  #results in USBTimeoutError
  File "C:\Users\kluges1\AppData\Local\Programs\Python\Python310\lib\site-packages\usb\core.py", line 1029, in read
    ret = fn(
  File "C:\Users\kluges1\AppData\Local\Programs\Python\Python310\lib\site-packages\usb\backend\libusb0.py", line 553, in bulk_read
    return self.__read(_lib.usb_bulk_read,
  File "C:\Users\kluges1\AppData\Local\Programs\Python\Python310\lib\site-packages\usb\backend\libusb0.py", line 692, in __read
    ret = int(_check(fn(
  File "C:\Users\kluges1\AppData\Local\Programs\Python\Python310\lib\site-packages\usb\backend\libusb0.py", line 447, in _check
    raise USBError(errmsg, ret)
usb.core.USBError: [Errno None] b'libusb0-dll:err [_usb_reap_async] timeout error\n'
PS C:\Users\kluges1\Downloads\rasodalo> 

```

**Comment by jonasmalacofilho** [comment, marked as answer]
I'm not familiar with that Decibel Meter, and also not sure what exactly you want to achieve, but here are some general notes...

> ```
> pysub               0.2
> pyusb               1.2.1
> ```

Probably unrelated to your question, but having two PyUSB versions in the same environment is likely to cause issues at some point.

> I have used Zadig to replace the drivers with libUSB-win32.

Why have you picked that driver?  I generally recommend starting with WinUSB.

On a related note, I would also recommend starting with the libusb-1.0 library and backend, instead of libusb0.

Other drivers and/or backends can sometimes be necessary, but they should be choices of last resort, as they are usually unmaintained.

> ```
>  iManufacturer          :    0x1 Silicon Labs
>  iProduct               :    0x2 CP2104 USB to UART Bridge Controller
>   [ ... ]
>     iInterface         :    0x2 CP2104 USB to UART Bridge Controller
> ```

At the USB level that device is a USB to UART bridge; with the stock driver, it should expose a serial port.  So an alternative approach might be to stick with the stock SiUSBx (IIRC) driver and use something like [pySerial](https://pyserial.readthedocs.io).

Still, similar bridges are known to work with PyUSB, using the WinUSB driver and libusb-1.0, and as long as you have implemented the protocol correctly.  In particular, you'll need some magic control transfers; if you don't have documentation, you'll need to match what you see in USB traffic captures and/or read some GPL-licensed code.

**Reply by ckl67** [reply]
Hello Jonas,

Thank you very much for your prompt answer.
I will try to follow your advice and revert back.
Fyi, I could log the exchanges between the device and the host thanks to WhireShark .
And even it is in UART mode, it seems that it follows a Bulk exchange.

I have given more information through 

https://stackoverflow.com/questions/73031781/ho-to-send-a-bulk-in-message-with-pyusb-with-python

Regards 
Christian

**Reply by jonasmalacofilho** [reply]
> And even it is in UART mode, it seems that it follows a Bulk exchange.

Yes, it's a custom USB protocol, using control and bulk transfers, to create a bridge for UART: that is, a device that (with suitable kernel drivers) looks to your kernel like a serial port, but is connected through USB.

> https://stackoverflow.com/questions/73031781/ho-to-send-a-bulk-in-message-with-pyusb-with-python

The general answer to your question there is that you need to understand this custom protocol.[^1]

But, at a glance, have you tried simply reading from the bulk endpoint _after (successfully) mimicking the control transfer that precedes it?_

```py
dev.read(0x82, 13)
```

The "successfully" part is important: if the control transfer doesn't produce the same result, you'll probably need something else to get the bridge into a state you can do the bulk read.

[^1]: There could also be other stuff to consider, like, you could need to place the device in a specific page, or something like that.  So, in reality, it can depend on both the protocol of the bridge and the protocol used at the UART level, which could be non-trivial (for example, in the past I've had to deal with PMBus over USB).

**Reply by ManiDeepak49** [reply]
i have also used libusb0 as backend could help me to resolve this issue. Help me how to change the backend to WinUSB or libUsb1

**Comment by ckl67** [comment]
Again,
thank you very much  for your answer.
Unfortunately not working.
```

import logroll
import datetime
import time
import usb.core

USB_idVendor = 0x10C4
USB_idProduct = 0xEA60

def connect():
    dev = usb.core.find(idVendor=USB_idVendor, idProduct=USB_idProduct)
    assert dev is not None
    print(dev)
    return dev

if __name__ == "__main__":

    # connect Sound Level meter DEM202 over USB
    dev = connect()

    while True:
        now = datetime.datetime.now()

        ret = dev.ctrl_transfer(0xC0, 16, 0, 0, 200)
        print(ret)

        val = dev.read(0x82, 13)
        print(val)

        time.sleep(1)
```
Provides
```

python3 rasodalo.py 
DEVICE ID 10c4:ea60 on Bus 007 Address 002 =================
 bLength                :   0x12 (18 bytes)
 bDescriptorType        :    0x1 Device
 bcdUSB                 :  0x200 USB 2.0
 bDeviceClass           :    0x0 Specified at interface
 bDeviceSubClass        :    0x0
 bDeviceProtocol        :    0x0
 bMaxPacketSize0        :   0x40 (64 bytes)
 idVendor               : 0x10c4
 idProduct              : 0xea60
 bcdDevice              :  0x100 Device 1.0
 iManufacturer          :    0x1 Silicon Labs
 iProduct               :    0x2 CP2104 USB to UART Bridge Controller
 iSerialNumber          :    0x3 01DA0D3D
 bNumConfigurations     :    0x1
  CONFIGURATION 1: 100 mA ==================================
   bLength              :    0x9 (9 bytes)
   bDescriptorType      :    0x2 Configuration
   wTotalLength         :   0x20 (32 bytes)
   bNumInterfaces       :    0x1
   bConfigurationValue  :    0x1
   iConfiguration       :    0x0 
   bmAttributes         :   0x80 Bus Powered
   bMaxPower            :   0x32 (100 mA)
    INTERFACE 0: Vendor Specific ===========================
     bLength            :    0x9 (9 bytes)
     bDescriptorType    :    0x4 Interface
     bInterfaceNumber   :    0x0
     bAlternateSetting  :    0x0
     bNumEndpoints      :    0x2
     bInterfaceClass    :   0xff Vendor Specific
     bInterfaceSubClass :    0x0
     bInterfaceProtocol :    0x0
     iInterface         :    0x2 CP2104 USB to UART Bridge Controller
      ENDPOINT 0x1: Bulk OUT ===============================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :    0x1 OUT
       bmAttributes     :    0x2 Bulk
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0x0
      ENDPOINT 0x82: Bulk IN ===============================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :   0x82 IN
       bmAttributes     :    0x2 Bulk
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0x0
array('B', [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
Traceback (most recent call last):
  File "rasodalo.py", line 26, in <module>
    val = dev.read(0x82, 13)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/core.py", line 1021, in read
    intf, ep = self._ctx.setup_request(self, endpoint)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/core.py", line 113, in wrapper
    return f(self, *args, **kwargs)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/core.py", line 229, in setup_request
    self.managed_claim_interface(device, intf)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/core.py", line 113, in wrapper
    return f(self, *args, **kwargs)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/core.py", line 178, in managed_claim_interface
    self.backend.claim_interface(self.handle, i)
  File "/home/christian/.local/lib/python3.6/site-packages/usb/backend/libusb1.py", line 829, in claim_interface
    _check(self.lib.libusb_claim_interface(dev_handle.handle, intf))
  File "/home/christian/.local/lib/python3.6/site-packages/usb/backend/libusb1.py", line 604, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 16] Resource busy
```

The issue must be linked with the protocol, which would be very difficult to understand

**Reply by jonasmalacofilho** [reply]
> The issue must be linked with the protocol, which would be very difficult to understand

The CP2104 [datasheet](https://www.silabs.com/documents/public/data-sheets/cp2104.pdf) states (on p. 19) it as compatible with the USBXpress drivers, and references some relevant documentation about that protocol.  But note that's not the only possible driver and, possibly, protocol (so it might not match exactly what you see in the captured traffic).

And can you comply with the GPL?  If you can, you might also be able to learn a lot from existing projects:

- open implementations of USBXpress drivers:
    * https://github.com/craigshelley/SiUSBXp (http://www.etheus.net/SiUSBXp_Linux_Driver)
    * https://github.com/fmeow/silabs_usb_xpress
- additionally, the Linux kernel also has a CP21xx driver.
